database-plugins/src/main/java/co/cask/hydrator/plugin/db/batch/source/DBSource.java

```diff
@@ -66,6 +85,135 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
     dbManager.validateJDBCPluginPipeline(pipelineConfigurer, getJDBCPluginId());
   }
 
+  class ListTablesRequest {
+    public String connectionString;
+    @Nullable
+    public String user;
+    @Nullable
+    public String password;
+    public String jdbcPluginName;
+    @Nullable
+    public String jdbcPluginType = "jdbc";
+  }
+
+  class GetSchemaRequest extends ListTablesRequest {
+    public String importQuery;
+  }
+
+  /**
+   * Endpoint method to get the list of tables in a database
+   * @param request {@link ListTablesRequest} containing information required for connection
+   * @param pluginContext context to create plugins
+   * @return list of table names
+   * @throws SQLException
+   * @throws InstantiationException
+   * @throws IllegalAccessException
+   * @throws BadRequestException If executing get tables threw {@link SQLSyntaxErrorException}
+   * we get the message and throw it as BadRequestException
+   */
+  @Path("listTables")
+  public List<String> listTables(ListTablesRequest request,
+                                 EndpointPluginContext pluginContext) throws SQLException,
+    InstantiationException, IllegalAccessException, BadRequestException {
+
+    JDBCDriverShim driverShim = loadPluginClassAndGetDriver(request, pluginContext);
+    List<String> tableNames = new ArrayList();
+
+    try {
+      // initialize connection
+      try (Connection connection = getConnection(request.connectionString, request.user, request.password)) {
+        DatabaseMetaData metadata = connection.getMetaData();
+        try (ResultSet rs = metadata.getTables(null, null, "%", null)) {
+          while (rs.next()) {
+            // table name is the third column and column index starts from 1.
+            String tableName = rs.getString(3);
+            if (tableName != null) {
+              tableNames.add(tableName);
+            }
+          }
+        }
+      }
+      return tableNames;
+    } catch (SQLException e) {
+      LOG.error("Exception while trying to get the tables in a database for connection {}.",
+                request.connectionString, e);
+      if (e instanceof SQLSyntaxErrorException) {
+        throw new BadRequestException(e.getMessage());
+      }
+      throw Throwables.propagate(e);
+    } finally {
+      DBUtils.deregisterShim(driverShim);
+    }
+  }
+
+  /**
+   * Endpoint method to get the output schema of a query.
+   * @param request {@link GetSchemaRequest} containing information required for connection and query to execute.
+   * @param pluginContext context to create plugins
+   * @return Map of field name to field type.
+   * @throws SQLException
+   * @throws InstantiationException
+   * @throws IllegalAccessException
+   * @throws BadRequestException If executing the query threw {@link SQLSyntaxErrorException}
+   * we get the message and throw it as BadRequestException
+   */
+  @Path("getSchema")
+  public Map<String, String> getSchema(GetSchemaRequest request,
```

Comment:
this schema returned here is different than the schema we use everywhere else (avro style schema).  Can we just return a cdap Schema instead of a Map? Or Schema.toString() instead of the map? That way you can call this method, then plug it directly into a schema field.